README.md
# Multfly pseudo-random number generator
NOTE: it is still in very early stage of development.

This repository contains reference and optimized implementations of Multfly pseudo-random number generator. Their API is intended to be minimal and not much end-user friendly.

## Why do you develop it?
My primary motivation is to make a pseudo-random number generator (PRNG) suitable for deep neural network training. Though speed requirement is modest compared to Monte-Carlo simulation, there are some challenges.

* Massively parallel execution: the largest GPGPU supercomputers consist of millions of CUDA cores. Each CUDA core will execute logically different threads (arranged as grids), so the total number of threads can be in the order of trillion. Bookkeeping appropriate number of PRNG instances is not trivial task.
* Heterogeneous execution: in deep neural network training, it is not only GPUs who need random numbers. In typical case, CPUs also relies on them for dataset sampling and data augmentation. You might want to move data augmentation routines to GPUs for performance. But how to preserve determinism?
* Importance of determinism: whereas determinism given by seed benefits every randomized program, in machine learning domain it is especially beneficial. Performance of some algorithm on some data can heavily depend on random sampling results. Since we couldn't know in advance what algorithm and data have such property, we should be conservative.
* Non-deterministic scheduling: it is not fixed what processor is assigned to each data. How to generate the same number sequence regardless of scheduling? Shared PRNG states come at a cost.
* Memory heavy: except convolution and matrix product, current deep neural networks are bottlenecked by memory bandwidth. Well, there is a good news: we can reduce many memory accesses by eliminating temporary arrays (loop fusion.) PRNG should be friendly to such optimization.
* Development model: deep neural networks are rarely implemented by numerical computation experts. So API must be easy to use.

There are shortage in existing designs.

* Linear feedback shift register, multiple recursive generators, and linear congruential generator: those are not perfectly suitable for parallel execution. Even though they have "jumpahead" functionality, jump distance and computation parallelism must be carefully chosen for performance. There are libraries that achieves admirable performance, but relying on them often introduce temporal arrays. They also tend to have distinguishable patterns.
* Counter-based PRNG: actually it is based on block cipher. It accepts k-bit key and k-bit counter, and generates k-bit output. The reasonable choice would be k=128, as 128-bit counter is enough for decades... It's true if users wisely split the 128-bit counter space, e.g. 32 bits to count iterations, 16 bits to number each layer, 16 bits for rejection sampling and the other 64 bits for parallelism. But such splitting requires users' effort. I don't want to impose such a non-essential effort. Generating 128-bit keys by a hash function risks collision. Again, I don't want to force users not to generate 2^64 keys for decades. Setting k=256 slows down everything, despite users often need just 32 bits.

## Install
Just copy `include/*` to `/usr/local/include` or somewhere.

## Usage
Currently there is no documentation. See `tests/*`.

## Build
`make all` builds test commands. You need [TestU01 1.23](http://simul.iro.umontreal.ca/testu01/tu01.html) to build BigCrush. You need CUDA to build `test/test_cuda`.

## Performance
I'll add benchmark result. Expected to be fast on GPU and multi-core SIMD CPU. Sequential implementation should be not much slow.

## Design
* 256-bit key and 128-bit counter, generating 128-bit number
* Efficient in 4-way SIMD
* The counter is split into 64-bit thread index, 32-bit sequential counter and 32-bit auxiliary bits
* ARX-like network
* Combined with multiplication and butterfly-indexing like communication, which is origin of the name Multfly
* Not taking high 32-bit of multiplication to avoid performance drop on some architectures (e.g. NVidia GPU after Maxwell gen.)
* Keys are generated from 256-bit name and 128-bit user value
* I intend that a unique name is set to every neural net layer.
* The user value is split into 64-bit global counter, which can be e.g. iteration number, and 64-bit global seed
* Key generation is not slow on any platform, thanks to Chacha8 cipher
* Key splitting is also supported, because we wouldn't like to name everywhere we generate random numbers

## TODO
* DNN training example using Chainer, CuPy and Cython (planed to be on another repository)
* benchmark
* algorithm fix
* more documentation
* AVX2 implementation
* API fix


CHANGELOG.md
# 0.0.1
* Initial version
* Very early stage development
* Tested reference, SSE4.1 and CUDA implementation
* Anything is not fixed


test/test.c
#include <inttypes.h>
#include <stddef.h>
#include <stdio.h>
#include <string.h>

#ifdef SSE41
#include <multfly_sse41.h>
#else
#include <multfly_ref.h>
#endif

int check_number32(char *title, uint32_t result, uint32_t expected) {
	puts(title);
	if (result == expected) {
		puts("ok");
		return 0;
	} else {
		printf("Result: 0x%08"PRIX32", Expected: 0x%08"PRIX32"\n", result, expected);
		return 1;
	}
}

int check_number64(char *title, uint64_t result, uint64_t expected) {
	puts(title);
	if (result == expected) {
		puts("ok");
		return 0;
	} else {
		printf("Result: 0x%016"PRIX64", Expected: 0x%016"PRIX64"\n", result, expected);
		return 1;
	}
}

int test_init() {
	int ng = 0;

	multfly_key keyn;
	multfly_initkey(&keyn, NULL, 0, 0);
	ng += check_number32("test_init NULL, 0, 0 [0]", keyn.v_[0], UINT32_C(0x5A7F760C));
	ng += check_number32("test_init NULL, 0, 0 [1]", keyn.v_[1], UINT32_C(0xE3ABAB0B));
	ng += check_number32("test_init NULL, 0, 0 [2]", keyn.v_[2], UINT32_C(0xECBCFCC9));
	ng += check_number32("test_init NULL, 0, 0 [3]", keyn.v_[3], UINT32_C(0x7CD8982B));
	ng += check_number32("test_init NULL, 0, 0 [4]", keyn.v_[4], UINT32_C(0xE428B5C8));
	ng += check_number32("test_init NULL, 0, 0 [5]", keyn.v_[5], UINT32_C(0xA1C1A56E));
	ng += check_number32("test_init NULL, 0, 0 [6]", keyn.v_[6], UINT32_C(0x0A66CADF));
	ng += check_number32("test_init NULL, 0, 0 [7]", keyn.v_[7], UINT32_C(0x16926033));

	multfly_name name0 = {0};
	multfly_key key0;
	multfly_initkey(&key0, &name0, 0, 0);
	ng += check_number32("test_init {0}, 0, 0 [0]", key0.v_[0], keyn.v_[0]);
	ng += check_number32("test_init {0}, 0, 0 [1]", key0.v_[1], keyn.v_[1]);
	ng += check_number32("test_init {0}, 0, 0 [2]", key0.v_[2], keyn.v_[2]);
	ng += check_number32("test_init {0}, 0, 0 [3]", key0.v_[3], keyn.v_[3]);
	ng += check_number32("test_init {0}, 0, 0 [4]", key0.v_[4], keyn.v_[4]);
	ng += check_number32("test_init {0}, 0, 0 [5]", key0.v_[5], keyn.v_[5]);
	ng += check_number32("test_init {0}, 0, 0 [6]", key0.v_[6], keyn.v_[6]);
	ng += check_number32("test_init {0}, 0, 0 [7]", key0.v_[7], keyn.v_[7]);

	multfly_key key;
	multfly_initkey_fromliteral(&key, "abcdefghijklmnopqrstuvwxyz123456", 78, 90);
	ng += check_number32("test_init abcdefghijklmnopqrstuvwxyz123456, 78, 90 [0]", key.v_[0], UINT32_C(0x1FCE68FA));
	ng += check_number32("test_init abcdefghijklmnopqrstuvwxyz123456, 78, 90 [1]", key.v_[1], UINT32_C(0x01D594AD));
	ng += check_number32("test_init abcdefghijklmnopqrstuvwxyz123456, 78, 90 [2]", key.v_[2], UINT32_C(0x8FEC7E65));
	ng += check_number32("test_init abcdefghijklmnopqrstuvwxyz123456, 78, 90 [3]", key.v_[3], UINT32_C(0x0FFC302B));
	ng += check_number32("test_init abcdefghijklmnopqrstuvwxyz123456, 78, 90 [4]", key.v_[4], UINT32_C(0x0201AC2E));
	ng += check_number32("test_init abcdefghijklmnopqrstuvwxyz123456, 78, 90 [5]", key.v_[5], UINT32_C(0x5DE5F28B));
	ng += check_number32("test_init abcdefghijklmnopqrstuvwxyz123456, 78, 90 [6]", key.v_[6], UINT32_C(0xD44262CC));
	ng += check_number32("test_init abcdefghijklmnopqrstuvwxyz123456, 78, 90 [7]", key.v_[7], UINT32_C(0xA0F8A154));

	return ng;
}

int test_gen32() {
	int ng = 0;
	multfly_key key;
	multfly_initkey(&key, NULL, 0, 0);
	uint32_t result[4];

	multfly_gen32(&key, 0, 0, result);
	ng += check_number32("test_gen32 [0]", result[0], UINT32_C(0x28DAC85C));
	ng += check_number32("test_gen32 [1]", result[1], UINT32_C(0x0D05143D));
	ng += check_number32("test_gen32 [2]", result[2], UINT32_C(0xF13F16D2));
	ng += check_number32("test_gen32 [3]", result[3], UINT32_C(0x5790CBA7));

	multfly_gen32(&key, 100, 0, result);
	ng += check_number32("test_gen32 [100]", result[0], UINT32_C(0xAC05F32E));
	ng += check_number32("test_gen32 [101]", result[1], UINT32_C(0xCAF615DD));
	ng += check_number32("test_gen32 [102]", result[2], UINT32_C(0x81CCDA48));
	ng += check_number32("test_gen32 [103]", result[3], UINT32_C(0xDC5C5E1A));

	return ng;
}

int test_gen64() {
	int ng = 0;
	multfly_key key;
	multfly_initkey(&key, NULL, 0, 0);
	uint64_t result[4];

	multfly_gen64(&key, 0, 0, result);
	ng += check_number64("test_gen64 [0]", result[0], UINT64_C(0x1BDCBA9328DAC85C));
	ng += check_number64("test_gen64 [1]", result[1], UINT64_C(0x65985DB70D05143D));
	ng += check_number64("test_gen64 [2]", result[2], UINT64_C(0x483C27C0F13F16D2));
	ng += check_number64("test_gen64 [3]", result[3], UINT64_C(0xE764123A5790CBA7));

	multfly_gen64(&key, 100, 0, result);
	ng += check_number64("test_gen64 [100]", result[0], UINT64_C(0xDD5542F5AC05F32E));
	ng += check_number64("test_gen64 [101]", result[1], UINT64_C(0x5A190C48CAF615DD));
	ng += check_number64("test_gen64 [102]", result[2], UINT64_C(0x9AE4C7C581CCDA48));
	ng += check_number64("test_gen64 [103]", result[3], UINT64_C(0x27A5E2FCDC5C5E1A));

	return ng;
}

int test_genf32() {
	int ng = 0;
	multfly_key key;
	multfly_initkey(&key, NULL, 0, 0);

	uint32_t u32_result[4];
	float f32_result[4];
	for (int idx = 0; idx < 100; idx++) {
		multfly_gen32(&key, idx, 0, u32_result);
		multfly_genf32(&key, idx, 0, f32_result);
		for (int lane = 0; lane < 4; lane++) {
			if (f32_result[lane] != (float)(u32_result[lane] >> 8) * (float)(1.0 / (UINT32_C(1) << 24))) {
				ng += 1;
			}
		}
	}

	puts("f32 correctness");
	if (!ng) {
		puts("ok");
	} else {
		printf("%d f32 samples failed\n", ng);
	}
	return ng;
}

int test_genf64() {
	int ng = 0;
	multfly_key key;
	multfly_initkey(&key, NULL, 0, 0);

	uint64_t u64_result[4];
	double f64_result[4];
	for (int idx = 0; idx < 100; idx++) {
		multfly_gen64(&key, idx, 0, u64_result);
		multfly_genf64(&key, idx, 0, f64_result);
		for (int lane = 0; lane < 4; lane++) {
			if (f64_result[lane] != (u64_result[lane] >> 11) * (1.0 / (UINT64_C(1) << 53))) {
				ng += 1;
			}
		}
	}

	puts("f64 correctness");
	if (!ng) {
		puts("ok");
	} else {
		printf("%d f64 samples failed\n", ng);
	}
	return ng;
}

int main() {
	int ng = 0;
	ng += test_init();
	ng += test_gen32();
	ng += test_gen64();
	ng += test_genf32();
	ng += test_genf64();

	if (ng == 0) {
		puts("All tests succeeded.");
	} else {
		printf("%d tests failed.\n", ng);
	}
	return ng != 0;
}


test/bigcrush.c
#include <stddef.h>

#include <unif01.h>
#include <bbattery.h>

#include <multfly_ref.h>

uint32_t multfly() {
	static multfly_key key;
	static uint64_t idx = 0;
	static uint32_t buf[4];
	static int pos = 0;
	if (pos == 0) {
		multfly_initkey(&key, NULL, 0, 0);
		pos = 4;
	}
	if (pos == 4) {
		pos = 0;
#ifdef GRAYCODE
		uint64_t idx_passed = idx ^ (idx >> 1);
#else
		uint64_t idx_passed = idx;
#endif
		multfly_gen32(&key, idx_passed, 0, buf);
		idx++;
	}
	uint32_t r = buf[pos++];

#ifdef BITREV
	uint32_t t;
	t = r & UINT32_C(0x00FF00FF);
	r = (t << 16) | (t >> 16) | (t ^ r);
	t = r & UINT32_C(0x0F0F0F0F);
	r = (t << 8) | (t >> 24) | (t ^ r);
	t = r & UINT32_C(0x33333333);
	r = (t << 4) | (t >> 28) | (t ^ r);
	t = r & UINT32_C(0x55555555);
	r = (t << 2) | (t >> 30) | (t ^ r);
	r = (r << 1) | (r >> 31);
#endif
	return r;
}

int main() {
	unif01_Gen *gen;
	gen = unif01_CreateExternGenBits("multfly", multfly);
	bbattery_BigCrush(gen);
	unif01_DeleteExternGenBits(gen);
	return 0;
}


src/multfly.c
#include <multfly.h>

static uint32_t multfly_rotl(uint32_t k, int n) {
	return (k << n) | (k >> (32 - n));
}

static void multfly_chacha_qround(uint32_t v[16], int a, int b, int c, int d) {
	v[a] += v[b];
	v[d] = multfly_rotl(v[d] ^ v[a], 16);
	v[c] += v[d];
	v[b] = multfly_rotl(v[b] ^ v[c], 12);
	v[a] += v[b];
	v[d] = multfly_rotl(v[d] ^ v[a], 8);
	v[c] += v[d];
	v[b] = multfly_rotl(v[b] ^ v[c], 7);
}

static void multfly_chacha8_permute(uint32_t v[16]) {
	for (int i = 0; i < 8; i += 2) {
		multfly_chacha_qround(v, 0, 4,  8, 12);
		multfly_chacha_qround(v, 1, 5,  9, 13);
		multfly_chacha_qround(v, 2, 6, 10, 14);
		multfly_chacha_qround(v, 3, 7, 11, 15);
		multfly_chacha_qround(v, 0, 5, 10, 15);
		multfly_chacha_qround(v, 1, 6, 11, 12);
		multfly_chacha_qround(v, 2, 7,  8, 13);
		multfly_chacha_qround(v, 3, 4,  9, 14);
	}
}

static multfly_key multfly_init_impl(const uint32_t seed[8], uint64_t global_seed, uint64_t global_ctr, uint32_t residual[8]) {
	uint32_t v[16];
	// Use Chacha IV: "expand 32-byte k"
	v[0]  = UINT32_C(0x61707865);
	v[1]  = UINT32_C(0x3320646E);
	v[2]  = UINT32_C(0x79622D32);
	v[3]  = UINT32_C(0x6B206574);
	v[4]  = seed[0];
	v[5]  = seed[1];
	v[6]  = seed[2];
	v[7]  = seed[3];
	v[8]  = seed[4];
	v[9]  = seed[5];
	v[10] = seed[6];
	v[11] = seed[7];
	v[12] = (uint32_t)global_ctr;
	v[13] = (uint32_t)(global_ctr >> 32);
	v[14] = (uint32_t)global_seed;
	v[15] = (uint32_t)(global_seed >> 32);

	multfly_chacha8_permute(v);

	if (residual != 0) {
		for (int i = 0; i < 8; i++) {
			residual[i] = v[i];
		}
	}

	multfly_key newkey;
	for (int i = 0; i < 8; i++) {
		newkey.k[i] = v[i + 8];
	}
	return newkey;
}

multfly_key multfly_init(const uint32_t seed[8], uint64_t global_seed, uint64_t global_ctr) {
	return multfly_init_impl(seed, global_seed, global_ctr, 0);
}

multfly_key multfly_split(multfly_key *key) {
	return multfly_init_impl(&key->k[0], 0, 0, &key->k[0]);
}

static void multfly_gen_qround(uint32_t us[4], uint32_t vs[4], uint32_t i) {
	uint32_t mulu = UINT32_C(2718281829);
	uint32_t mulv = UINT32_C(3141592653);
	uint32_t incr = UINT32_C(0x33123456);

	uint32_t u = us[i];
	uint32_t v = vs[i];

	u += multfly_rotl(incr, i);
	v += u;
	v ^= multfly_rotl(u, 8);
	v *= mulv;
	u ^= multfly_rotl(v, 9);
	u += v;
	u *= mulu;
	v ^= multfly_rotl(u, 16);
	v += u;

	us[i] = u;
	vs[i] = v;
}

static void multfly_gen_swap(uint32_t us[4], uint32_t i, uint32_t j) {
	uint32_t tmp = us[i];
	us[i] = us[j];
	us[j] = tmp;
}

static void multfly_gen_impl(const multfly_key *key, uint64_t ctr, uint32_t result[4]) {
	uint32_t us[4];
	uint32_t vs[4];

	for (int i = 0; i < 4; i++) {
		us[i] = key->k[i] ^ (uint32_t)ctr;
		vs[i] = key->k[i + 4] ^ (uint32_t)(ctr >> 32);
		++ctr;
	}

	multfly_gen_qround(us, vs, 0);
	multfly_gen_qround(us, vs, 1);
	multfly_gen_qround(us, vs, 2);
	multfly_gen_qround(us, vs, 3);

	multfly_gen_swap(us, 0, 1);
	multfly_gen_swap(us, 2, 3);

	multfly_gen_qround(us, vs, 0);
	multfly_gen_qround(us, vs, 1);
	multfly_gen_qround(us, vs, 2);
	multfly_gen_qround(us, vs, 3);

	multfly_gen_swap(us, 0, 2);
	multfly_gen_swap(us, 1, 3);

	multfly_gen_qround(us, vs, 0);
	multfly_gen_qround(us, vs, 1);
	multfly_gen_qround(us, vs, 2);
	multfly_gen_qround(us, vs, 3);

	result[0] = vs[0];
	result[1] = vs[1];
	result[2] = vs[2];
	result[3] = vs[3];
}

static void multfly_gen_loop(const multfly_key *key, uint64_t ctr, uint32_t *array, uint64_t len) {
	for (uint64_t i = 0; i < len; i += 4, ctr += 4) {
		multfly_gen_impl(key, ctr, array + i);
	}
}

void multfly_gen(const multfly_key *key, uint64_t ctr, uint32_t *array, uint64_t len) {
	ctr &= -4;
	len &= -4;

	multfly_gen_loop(key, ctr, array, len);
}

void multfly_gen128(const multfly_key *key, uint64_t ctr, uint32_t *array[4], uint64_t len) {
	ctr &= -4;
	len &= -4;

	multfly_gen_loop(key, ctr + 0, array[0], len);
	multfly_gen_loop(key, ctr + 1, array[1], len);
	multfly_gen_loop(key, ctr + 2, array[2], len);
	multfly_gen_loop(key, ctr + 3, array[3], len);
}


src/multfly_ref.c
#include <multfly.h>

static uint32_t multfly_rotl(uint32_t k, int n) {
	return (k << n) | (k >> (32 - n));
}

static void multfly_chacha_qround(uint32_t v[16], int a, int b, int c, int d) {
	v[a] += v[b];
	v[d] = multfly_rotl(v[d] ^ v[a], 16);
	v[c] += v[d];
	v[b] = multfly_rotl(v[b] ^ v[c], 12);
	v[a] += v[b];
	v[d] = multfly_rotl(v[d] ^ v[a], 8);
	v[c] += v[d];
	v[b] = multfly_rotl(v[b] ^ v[c], 7);
}

static void multfly_chacha8_permute(uint32_t v[16]) {
	for (int i = 0; i < 8; i += 2) {
		multfly_chacha_qround(v, 0, 4,  8, 12);
		multfly_chacha_qround(v, 1, 5,  9, 13);
		multfly_chacha_qround(v, 2, 6, 10, 14);
		multfly_chacha_qround(v, 3, 7, 11, 15);
		multfly_chacha_qround(v, 0, 5, 10, 15);
		multfly_chacha_qround(v, 1, 6, 11, 12);
		multfly_chacha_qround(v, 2, 7,  8, 13);
		multfly_chacha_qround(v, 3, 4,  9, 14);
	}
}

static multfly_key multfly_derive_impl(const multfly_key *key, uint64_t global_seed, uint64_t global_ctr, multfly_key *splitkey) {
	uint32_t v[16];
	// Use Chacha IV: "expand 32-byte k"
	v[0]  = UINT32_C(0x61707865);
	v[1]  = UINT32_C(0x3320646E);
	v[2]  = UINT32_C(0x79622D32);
	v[3]  = UINT32_C(0x6B206574);
	v[4]  = key->k[0];
	v[5]  = key->k[1];
	v[6]  = key->k[2];
	v[7]  = key->k[3];
	v[8]  = key->k[4];
	v[9]  = key->k[5];
	v[10] = key->k[6];
	v[11] = key->k[7];
	v[12] = (uint32_t)global_ctr;
	v[13] = (uint32_t)(global_ctr >> 32);
	v[14] = (uint32_t)global_seed;
	v[15] = (uint32_t)(global_seed >> 32);

	multfly_chacha8_permute(v);

	multfly_key newkey;
	for (int i = 0; i < 8; i++) {
		newkey.k[i] = v[i];
	}
	if (splitkey != 0) {
		for (int i = 0; i < 8; i++) {
			splitkey->k[i] = v[i + 8];
		}
	}
	return newkey;
}

multfly_key multfly_derive(const multfly_key *key, uint64_t global_seed, uint64_t global_ctr) {
	return multfly_derive_impl(key, global_seed, global_ctr, 0);
}

multfly_key multfly_split(multfly_key *key) {
	return multfly_derive_impl(key, 0, 0, key);
}

static void multfly_gen_qround(uint32_t us[4], uint32_t vs[4], uint32_t i) {
	uint32_t mulu = UINT32_C(2718281829);
	uint32_t mulv = UINT32_C(3141592653);
	uint32_t incr = UINT32_C(0x33123456);

	uint32_t u = us[i];
	uint32_t v = vs[i];

	u += multfly_rotl(incr, i);
	v += u;
	v ^= multfly_rotl(u, 8);
	v *= mulv;
	u ^= multfly_rotl(v, 9);
	u += v;
	u *= mulu;
	v ^= multfly_rotl(u, 16);
	v += u;

	us[i] = u;
	vs[i] = v;
}

static void multfly_gen_swap(uint32_t us[4], uint32_t i, uint32_t j) {
	uint32_t tmp = us[i];
	us[i] = us[j];
	us[j] = tmp;
}

static void multfly_gen_impl(const multfly_key *key, uint64_t ctr, uint32_t result[4]) {
	uint32_t us[4];
	uint32_t vs[4];

	for (uint32_t i = 0; i < 4; i++) {
		us[i] = key->k[i] ^ ((uint32_t)ctr + i);
		vs[i] = key->k[i + 4] ^ ((uint32_t)(ctr >> 32) + i);
	}

	multfly_gen_qround(us, vs, 0);
	multfly_gen_qround(us, vs, 1);
	multfly_gen_qround(us, vs, 2);
	multfly_gen_qround(us, vs, 3);

	multfly_gen_swap(vs, 0, 1);
	multfly_gen_swap(vs, 2, 3);

	multfly_gen_qround(us, vs, 0);
	multfly_gen_qround(us, vs, 1);
	multfly_gen_qround(us, vs, 2);
	multfly_gen_qround(us, vs, 3);

	multfly_gen_swap(vs, 0, 2);
	multfly_gen_swap(vs, 1, 3);

	multfly_gen_qround(us, vs, 0);
	multfly_gen_qround(us, vs, 1);
	multfly_gen_qround(us, vs, 2);
	multfly_gen_qround(us, vs, 3);

	result[0] = vs[0];
	result[1] = vs[1];
	result[2] = vs[2];
	result[3] = vs[3];
}

static void multfly_gen_loop(const multfly_key *key, uint64_t ctr, uint32_t *array, uint64_t len) {
	for (uint64_t i = 0; i < len; i += 4, ctr += 4) {
		multfly_gen_impl(key, ctr, array + i);
	}
}

void multfly_gen(const multfly_key *key, uint64_t ctr, uint32_t *array, uint64_t len) {
	ctr &= -4;
	len &= -4;

	multfly_gen_loop(key, ctr, array, len);
}

void multfly_gen128(const multfly_key *key, uint64_t ctr, uint32_t *array[4], uint64_t len) {
	ctr &= -4;
	len &= -4;

	multfly_gen_loop(key, ctr + 0, array[0], len);
	multfly_gen_loop(key, ctr + 1, array[1], len);
	multfly_gen_loop(key, ctr + 2, array[2], len);
	multfly_gen_loop(key, ctr + 3, array[3], len);
}


include/multfly_types.h
#ifndef MULTFLY_MULTFLY_TYPES_H_
#define MULTFLY_MULTFLY_TYPES_H_

#ifdef __CUDACC_RTC__
// stdint.h doesn't work with NVRTC.
typedef unsigned int uint32_t;
typedef unsigned long long uint64_t;
#else
#include <stdint.h>
#endif

#ifdef __cplusplus
extern "C" {
#endif

typedef struct multfly_name_ {
	uint32_t v[8];
} multfly_name;

typedef struct multfly_key_ {
	uint32_t v_[8];
} multfly_key;

#ifdef __cplusplus
#define MULTFLY_NAME_CONSTRUCT_(...) static_cast<const multfly_name &>(multfly_name __VA_ARGS__)
#else
#define MULTFLY_NAME_CONSTRUCT_(...) ((multfly_name) __VA_ARGS__)
#endif

#define MULTFLY_NAME_LITERAL(literal) MULTFLY_NAME_CONSTRUCT_({{ \
	(uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[-(sizeof(literal) > 33)] /* Literal MUST NOT have more than 32 letters. */ | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[1] << 8) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[2] << 16) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[3] << 24), \
	(uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[4] | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[5] << 8) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[6] << 16) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[7] << 24), \
	(uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[8] | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[9] << 8) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[10] << 16) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[11] << 24), \
	(uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[12] | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[13] << 8) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[14] << 16) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[15] << 24), \
	(uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[16] | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[17] << 8) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[18] << 16) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[19] << 24), \
	(uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[20] | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[21] << 8) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[22] << 16) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[23] << 24), \
	(uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[24] | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[25] << 8) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[26] << 16) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[27] << 24), \
	(uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[28] | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[29] << 8) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[30] << 16) | \
	((uint32_t)(literal "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0")[31] << 24) \
}})

#ifdef __cplusplus
}
#endif

#endif
